Add CI tests for signed images plus improvements

rborn-tx opened the pull request:
This includes:

- A series of improvements to the script responsible for fetching Tezi images, adding flexibility and the ability to fetch images from the internal Artifactory repository.
- New test jobs for running "host-only" tests with signed images.
- Other minor changes.

Please check the commit messages for details.

Related-to: TCB-756, TCB-740

rborn-tx committed:
tests/integration: Normalize value/usage of IS_WIC variable

Originally:

- An environment variable called "IS_WIC" was set in the .gitlab-ci.yaml
  file and used by run_all.sh; form the perspective of that script, the
  allowed values were "true" or "false" (though anything other than
  "true" would be regarded as "false").
- The run_all.sh script would invoke run.sh which in turn would set a
  variable with the same name, but now with value "1" if the machine was
  listed as one of those using WIC images. Otherwise, it wouldn't set
  the variable.
- Therefore, from the perspective of the run.sh script, the variable
  could assume values "1", "0", "true" or "false", but the script would
  only handle the case where its value is "1".

With this commit:

- The variable must be set to "0" or "1" like all other "binary"
  variables used by the test scripts.
- run_all.sh will use the value to download the appropriate test
  images (WIC/non-WIC).
- run.sh will use the value to select the appropriate tests; if passed
  to the script, the value is honored; when not passed, it selects a
  default value based on the machine.

Overall, the idea is to make the handling of the variable more
consistent across the scripts.

Signed-off-by: Rogerio Guerra Borin <rogerio.borin@toradex.com>

rborn-tx committed:
gitlab-ci: Simplify references to the TCB image under test

This is basically a readability improvement for the gitlab-ci file:

- Export an environment variable (named TCB_IMAGE_UNDER_TEST) containing
  the full name of the TCB Docker image under test.
- Use that variable in all test jobs.

Signed-off-by: Rogerio Guerra Borin <rogerio.borin@toradex.com>

rborn-tx committed:
tests/integration: Allow fetching signed images plus improvements

Add a series of improvements to "get_tezi_images.sh", the image fetching
script:

- Allow the Artifactory URLs to be specified (overridden) through env
  variables.

- Allow the acceptable distros to be specified, so only those will be
  considered for download (e.g. "torizon torizon-upstream").

- Allow the acceptable variants to be specified, so only those will be
  considered for download (e.g. "torizon-docker torizon-minimal").

- Add support for fetching signed images; at the time of this writing
  those are only available on the internal Artifactory which now can be
  used thanks to the above mentioned improvements.

- Do a deep search rather than shallow search when looking for images.
  Originally the latest build for a machine was found where the images
  where looked for. Now scripts goes over all builds/distros/variants
  untils in finds the specified combination.

Related-to: TCB-756, TCB-740

Signed-off-by: Rogerio Guerra Borin <rogerio.borin@toradex.com>

rborn-tx committed:
tests/integration: Add tests for tdx-signed images

Add jobs for testing TorizonCore Builder with tdx-signed images; only
"host" test cases are run for now.

Related-to: TCB-756

Signed-off-by: Rogerio Guerra Borin <rogerio.borin@toradex.com>

rborn-tx committed:
torizoncore-builder: Fix exception field access

Check the "output" field of an exception object before accessing its
contents. In some cases, an exception may be thrown with that field not
set which would cause a traceback to be dumped besides the actual error
message. This was detected when testing "filesystem" changes with the
.tcattr file having files not actually present in the changes directory.

Here's a stretch of the build command output showing that situation:

  =>> Handling output section
  setfacl: shadow: No such file or directory
  setfacl: mtab: No such file or directory
  setfacl: locale.conf: No such file or directory
  setfacl: vconsole.conf: No such file or directory
  Command '['setfacl', ...]' returned non-zero exit status 1.

  Traceback (most recent call last):
    File "/builder/torizoncore-builder", line 232, in <module>
      mainargs.func(mainargs)
  ...
  During handling of the above exception, another exception occurred:

  Traceback (most recent call last):
    File "/builder/torizoncore-builder", line 243, in <module>
      if b"\n" in ex.output:
  TypeError: argument of type 'NoneType' is not iterable

Signed-off-by: Rogerio Guerra Borin <rogerio.borin@toradex.com>

rborn-tx commented:
Guys, please feel free to review this. I'm leaving it as draft to wait for https://github.com/torizon/torizoncore-builder/pull/97 to be merged. If we merge before it, the new test jobs would most certainly fail.

jsrc27 reviewed (approved):
LGTM

lucas-p-bernardes commented on .gitlab-ci.yml:
```diff
@@ -306,21 +315,63 @@ test-host-tos7-64bit-nightly:
 
+test-host-tos7-32bit-signed-nightly:
+  # TODO: Review once signed images are available publicly.
+  tags:
+    - torizoncore
```
Hey @rborn-tx, In the TCB pipelines we are using the "all" tag for jobs that access internal Toradex network resources, and no tag definition for jobs that don’t need internal resources (so they use the default tag "saas-linux-small-amd64").
The "all" tag is present on all internal runners, but the "torizoncore" tag is only present on the Horw1 runner. Is there any specific configuration on that runner that makes it the only suitable one for these jobs, or would the "all" tag already be adequate?

rborn-tx replied:
@lucasbernardestoradex I've switched to tag "all". Thanks for letting me know.

lucas-p-bernardes commented on .gitlab-ci.yml:
```diff
@@ -306,21 +315,63 @@ test-host-tos7-64bit-nightly:
+
+test-host-tos7-64bit-signed-nightly:
+  # TODO: Review once signed images are available publicly.
+  tags:
+    - torizoncore
```
Same here!

rborn-tx replied:
Ditto

lucas-p-bernardes commented on tests/integration/get_tezi_images.sh:
```diff
@@ -1,86 +1,132 @@
+    for build in ${build_numbers}; do
+        machine_url="${artifactory_url}/${build}/${TCB_MACHINE}/"
+        dbg "Accessing ${machine_url}"
+        if ! wget --spider -q "${machine_url}"; then
+            log "Build #${build}: not available for machine ${TCB_MACHINE}"
+            continue
```
Hi @rborn-tx, based on the commit message description below and this part of the code, it looks like you fixed bug TCB-740 here. Could you please add the "Related-to: TCB-740" in the commit and log that this has already been done in the ticket?

```
tests/integration: Allow fetching signed images plus improvements
[...]
- Do a deep search rather than shallow search when looking for images.
  Originally the latest build for a machine was found where the images
  where looked for. Now scripts goes over all builds/distros/variants
  untils in finds the specified combination.
```

rborn-tx replied:
Done!

lucas-p-bernardes reviewed (approved):
LGTM! Very good improvements!

rborn-tx committed:
tests/integration: Adapt some tests for signed images

Adapt some of the "build" command tests for signed images, more
precisely for images having a kernel in the FIT format.

Related-to: TCB-756

Signed-off-by: Rogerio Guerra Borin <rogerio.borin@toradex.com>

rborn-tx commented:
@lucas-akira Since the other reviewers have already approved this while in draft state, allow me to let you know that this is now ready for review. The new jobs that test with a signed image have run successfully on GitLab too (though there were errors in AVAL unrelated to signed images which I don't think block this PR or even our upcoming release).

lucas-akira reviewed (approved):
LGTM. Nice improvement to the image fetching script!
